### ISSUE TITLE
[Question] Is it possible to share an instantiated class between `globalSetup` and tests?

### ISSUE THREAD

**Issue by liamdocherty10** [microsoft/playwright#16999]
Following on from my question in #16800, is it possible to share an instantiated class between `globalSetup` and the tests?

My use case is to be able open a database connection in `globalSetup` and then be able to run queries as and when desired throughout my test files.

Without this functionality I am limited to running the test files sequentially because I can only connect to the DB on a single port. If each file creates it's own connection and the tests run in parallel I get an 'address already in use' error

I am aware that simple variables can be shared using environment variables.

E.g. something like:

- Global Setup

```ts
async function globalSetup(): Promise<void> {
  const dbConnector: DbConnector = new DbConnector();
  await dbConnector.connect();
}
```

- Test File:
```ts
import { dbConnector } from ...

test("My Test", async () => {
  const results = await dbConnector.executeQuery("SELECT * FROM ...")
});
```

**Comment by pavelfeldman** [comment]
No, this is not possible, global setup and your tests run in different operating system processes. You should open your db connection in either beforeAll hook or in the worker-level fixture of your tests.

**Comment by liamdocherty10** [comment]
Since I am limited to running the test files sequentially, I did try using `test.describe.parallel` + `test.beforeAll` + multiple workers, however because `beforeAll` runs once for each worker I still end up creating multiple connections and getting 'address already in use'  errors.

I also tried using an async describe but found that is unsupported too e.g.
```ts
test.describe("", async () => {
  const dbConnector: DbConnector = new DbConnector();
  await dbConnector.connect();

  test("", async () => {
    const results = await dbConnector.executeQuery("SELECT * FROM ...")
  });

  test("", async () => {
    const results = await dbConnector.executeQuery("SELECT * FROM ...")
  });
});
```

Do you have an example of how a worker-level fixture could be used instead?

**Comment by pavelfeldman** [comment]
Worker-level fixture will still create as many connections as there are workers. The idea of parallelization is that we don't share any data between processes, so you should create as many isolated connections as there are workers. Note that you can't run code in the describe block, it is not a program, it is a declaration of the suite. You can use beforeAll hook in those blocks though.

**Comment by liamdocherty10** [comment]
Ok so seems like my desired approach is simply unsupported, thanks for the reply.

Is there an open feature request for sharing an object between `globalSetup` and tests?

As an aside how do you see `beforeAll` and `beforeEach` differing in `test.describe.parallel`? They seem to act in the same way from what I've seen.

**Comment by pavelfeldman** [comment]
> Is there an open feature request for sharing an object between globalSetup and tests?

I don't think so, but this would not be possible due to multi-process architecture of Playwright. We can't have that if we want to be able to run tests in parallel for obvious reasons.

> As an aside how do you see beforeAll and beforeEach differing in test.describe.parallel? They seem to act in the same way from what I've seen.

`beforeAll` will run once before all the tests worker executes, `beforeEach` will run before each test worker executes. So yes, it is a bit confusing since `beforeAll` can run more than once for each file. Still it should only run once in each worker. You can verify this via passing `--workers=1`.

**Comment by pavelfeldman** [comment]
Closing as per above, please feel free to open a new issue if this does not cover your use case.
